src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -726,13 +726,37 @@ static int get_pool_offset(sd_dhcp_server *server, be32_t requested_ip) {
         return be32toh(requested_ip & ~server->netmask) - server->pool_offset;
 }
 
+static int prepare_new_lease(DHCPLease **ret_lease, be32_t address, DHCPClientId *client_id, uint8_t *chaddr, be32_t gateway, usec_t expiration){
+        DHCPLease *lease = new0(DHCPLease, 1);
+        if (!lease)
+                return -ENOMEM;
+
+        *lease = (DHCPLease){
+                .address = address,
+                .client_id.length = client_id->length,
+                .gateway = gateway,
+                .expiration = expiration
+        };
+        lease->client_id.data = memdup(client_id->data, client_id->length);
+        if (!lease->client_id.data) {
+                free(lease);
```

Comment:
instead of manually freeing things in the failure path here, please use `_cleanup_(sd_dhcp_lease_unrefp)` on the variable declaration, so that it is automatically freed when the function is left, except when assigned NULL, the way TAKE_PTR() does it beow.